kv/src/main/java/com/test/kv/lib/storage/Memory.java

```diff
@@ -0,0 +1,197 @@
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import com.sun.org.apache.bcel.internal.generic.INSTANCEOF;
+import com.sun.tools.jdi.ConcreteMethodImpl;
+import com.test.kv.exceptions.KVKeyDoesNotExistsException;
+import com.test.kv.interfaces.Storage;
+
+/**
+ * Implementation of Storage -- In memory store with LRU
+ */
+public class Memory<KeyType, ValueType> implements Storage<KeyType, ValueType> {
+    /**
+     * storage is the actual store where we keep all key->value pairs
+     */
+    protected Map<KeyType, ValueType> storage;
+
+    /**
+     * maxKeySize is the max number of keys storage can store
+     */
+    protected int maxKeySize = 1024;
+
+    /**
+     * expireAtStorage stores the key->TTL pair for cleanup expired keys usage
+     */
+    protected Map<KeyType, Long> expireAtStorage;
+
+    /**
+     * expiringQueue maintains the keys in a queue for LRU
+     */
+    protected ConcurrentLinkedQueue<KeyType> expiringQueue;
+
+    /**
+     * expiringQueueLock is a mutex that protects expiringQueue
+     */
+    protected static final Object expiringQueueLock = new Object();
+
+    /**
+     * static logger
+     */
+    private static final Logger log = Logger.getLogger(Memory.class.getName());
+
+    /**
+     * Constructor
+     */
+    public Memory() {
+        this(0); // If user doesn't specify cache size, use the default value
+    }
+
+    /**
+     * Constructor
+     *
+     * @param maxKeySize the maximum size of keys Memory can store
+     */
+    public Memory(int maxKeySize) {
+        if (this.maxKeySize != 0) {
+            this.maxKeySize = maxKeySize;
+        }
+        this.storage = new ConcurrentHashMap<KeyType, ValueType>(maxKeySize);
+        this.expireAtStorage = new ConcurrentHashMap<KeyType, Long>(maxKeySize);
+        this.expiringQueue = new ConcurrentLinkedQueue<KeyType>();
+    }
+
+    /**
+     * updateKeyRecentUsage updates the key in expiringQueue
+     *
+     * @param key the the to be updated in expiringQueue
+     */
+    private void updateKeyRecentUsage(KeyType key) {
+        synchronized (expiringQueueLock) {
+            expiringQueue.remove(key);
+            expiringQueue.offer(key);
+        }
+    }
+
+    /**
+     * removeLeastRecentUsedKey removes the key which is least recently used
+     */
+    private void removeLeastRecentUsedKey() {
+        synchronized (expiringQueueLock) {
+            KeyType key = expiringQueue.poll();
+            if (key != null) {
+                storage.remove(key);
+            }
+        }
+    }
+
+    /**
+     * storeKeyAndTTL stores the key with its ttl in expireAtStorage
+     */
+    private void storeKeyAndTTL(KeyType key, Long ttlInSec) {
+        Long expiredAt = new Long(Instant.now().getEpochSecond() + ttlInSec.longValue());
+        expireAtStorage.put(key, expiredAt);
```

Comment:
(trying to be very strict here, may not necessary in practice..)
since store need to touch 3 different data object without transaction protection, it got potential to result in ABA problem when frequently set and remove same key.